Errors in Kubernetes leads to other services down

Description:
Our Kubernetes cluster hosted on DigitalOcean is experiencing network connectivity issues, resulting in the disruption of several services within the cluster.

Symptoms:
The containers have Error status and didn't clean up automatically. Eventually, those not working and Error containers were using a lot of memory and cause the following problems.
1. Multiple services deployed on our Kubernetes cluster are unreachable or intermittently accessible.
2. Communication between pods appears to be sporadic or entirely unavailable.

Proposed Strategy for Resolution:
1. A straightforward approach involves considering a recycling strategy: removing error containers to prevent unnecessary memory growth.
2. We should review the configuration of Kubernetes and prioritize the pods.
3. We typically operate within the confines of utilizing only 80% of Kubernetes' memory capacity. There's potential to enhance our approach, such as by adjusting the machine configuration—potentially opting for either 2 nodes with 8GB memory each or an expansion to 4 nodes. Alternatively, we could consider isolating the monitoring pods onto a single node equipped with greater memory capacity. In any case, it's prudent to devise a more effective strategy for optimizing memory utilization.
![Screenshot 2023-08-18 at 3 09 04 AM](https://github.com/MurmurationsNetwork/MurmurationsServices/assets/8935531/b0aff25c-4b95-4a9f-ae6a-1fa4c8d00dcf)

Error logs for reference:
Logs from schemaparser.

```
goroutine 1 [running]:
go.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0x1?, 0xc000184500?, {0x0?, 0x0?, 0xc0001a2080?})
	/go/pkg/mod/go.uber.org/zap@v1.25.0/zapcore/entry.go:196 +0x65
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc000190820, {0xc0001a6540, 0x1, 0x1})
	/go/pkg/mod/go.uber.org/zap@v1.25.0/zapcore/entry.go:262 +0x3ec
go.uber.org/zap.(*Logger).Panic(0x8c9140?, {0x90d7c4?, 0xc000034048?}, {0xc0001a6540, 0x1, 0x1})
	/go/pkg/mod/go.uber.org/zap@v1.25.0/logger.go:258 +0x59
github.com/MurmurationsNetwork/MurmurationsServices/pkg/logger.Panic({0x90d7c4, 0x1f}, {0xa4fb80?, 0xc000184500?}, {0x0?, 0x0?, 0xc00002a060?})
	/src/schemaparser/pkg/logger/logger.go:79 +0x1b7
github.com/MurmurationsNetwork/MurmurationsServices/services/cronjob/schemaparser/pkg/schemaparser.NewCronJob()
	/src/schemaparser/services/cronjob/schemaparser/pkg/schemaparser/cronjob.go:24 +0xfc
main.main()
	/src/schemaparser/cmd/schemaparser/main.go:10 +0x31
{"level":"panic","time":"2023-08-18T06:05:19.212Z","msg":"error when trying to ping Redis","error":"dial tcp 10.245.169.146:6379: i/o timeout"}
```

logs from dataproxyupdater

```
{"level":"info","time":"2023-08-18T06:00:05.659Z","msg":"server selection error: context deadline exceeded, current topology: { Type: Unknown, Servers: [{ Addr: data-proxy-mongo:27017, Type: Unknown }, ] }, retry in 31 seconds \n"}
{"level":"info","time":"2023-08-18T06:00:38.609Z","msg":"server selection error: context deadline exceeded, current topology: { Type: Unknown, Servers: [{ Addr: data-proxy-mongo:27017, Type: Unknown, Last error: dial tcp 10.245.163.86:27017: i/o timeout }, ] }, retry in 68 seconds \n"}
{"level":"info","time":"2023-08-18T06:01:48.542Z","msg":"server selection error: context deadline exceeded, current topology: { Type: Unknown, Servers: [{ Addr: data-proxy-mongo:27017, Type: Unknown, Last error: dial tcp 10.245.163.86:27017: i/o timeout }, ] }, retry in 126 seconds \n"}
{"level":"info","time":"2023-08-18T06:03:56.467Z","msg":"server selection error: context deadline exceeded, current topology: { Type: Unknown, Servers: [{ Addr: data-proxy-mongo:27017, Type: Unknown, Last error: dial tcp 10.245.163.86:27017: i/o timeout }, ] }, retry in 236 seconds \n"}
{"level":"info","time":"2023-08-18T06:07:53.999Z","msg":"server selection error: context deadline exceeded, current topology: { Type: Unknown, Servers: [{ Addr: data-proxy-mongo:27017, Type: Unknown, Last error: dial tcp 10.245.163.86:27017: i/o timeout }, ] }, retry in 235 seconds \n"}
{"level":"panic","time":"2023-08-18T06:11:50.588Z","msg":"error when trying to ping the MongoDB","error":"server selection error: context deadline exceeded, current topology: { Type: Unknown, Servers: [{ Addr: data-proxy-mongo:27017, Type: Unknown, Last error: dial tcp 10.245.163.86:27017: i/o timeout }, ] }"}
panic: error when trying to ping the MongoDB

goroutine 1 [running]:
go.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0x1?, 0x42f8f0?, {0x0?, 0x0?, 0xc0000c3860?})
	/go/pkg/mod/go.uber.org/zap@v1.25.0/zapcore/entry.go:196 +0x65
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc000177110, {0xc000356000, 0x1, 0x1})
	/go/pkg/mod/go.uber.org/zap@v1.25.0/zapcore/entry.go:262 +0x3ec
go.uber.org/zap.(*Logger).Panic(0x8b2600?, {0x8eacfd?, 0x0?}, {0xc000356000, 0x1, 0x1})
	/go/pkg/mod/go.uber.org/zap@v1.25.0/logger.go:258 +0x59
github.com/MurmurationsNetwork/MurmurationsServices/pkg/logger.Panic({0x8eacfd, 0x25}, {0xa2bae0?, 0xc0002cbe50?}, {0x0?, 0x0?, 0x0?})
	/src/dataproxyupdater/pkg/logger/logger.go:79 +0x1b7
github.com/MurmurationsNetwork/MurmurationsServices/services/cronjob/dataproxyupdater/global.mongoInit()
	/src/dataproxyupdater/services/cronjob/dataproxyupdater/global/init.go:27 +0x165
github.com/MurmurationsNetwork/MurmurationsServices/services/cronjob/dataproxyupdater/global.Init()
	/src/dataproxyupdater/services/cronjob/dataproxyupdater/global/init.go:11 +0x1c
main.init.0()
	/src/dataproxyupdater/services/cronjob/dataproxyupdater/cmd/dataproxyupdater/main.go:26 +0x17
```

--- a/charts/murmurations/charts/dataproxyrefresher/templates/dataproxyrefresher/cronjob.yaml
+++ b/charts/murmurations/charts/dataproxyrefresher/templates/dataproxyrefresher/cronjob.yaml
@@ -10,8 +10,16 @@ spec:
   {{- else }}
   schedule: "*/2 * * * *"  # every 2 minutes
   {{- end }}
+  # Keep the latest successful job.
+  successfulJobsHistoryLimit: 1
+  # Keep the latest failed job.
+  failedJobsHistoryLimit: 1
   jobTemplate:
     spec:
+      # Retry the Job 3 times before marking it as failed.
+      backoffLimit: 3
+      # Delete the Job and its pods 24 hours after completion.
+      ttlSecondsAfterFinished: 86400
       template:
         metadata:
           annotations:

--- a/charts/murmurations/charts/dataproxyupdater/templates/dataproxyupdater/cronjob.yaml
+++ b/charts/murmurations/charts/dataproxyupdater/templates/dataproxyupdater/cronjob.yaml
@@ -8,8 +8,16 @@ spec:
   {{- else }}
   schedule: "*/2 * * * *"  # every 2 minutes
   {{- end }}
+  # Keep the latest successful job.
+  successfulJobsHistoryLimit: 1
+  # Keep the latest failed job.
+  failedJobsHistoryLimit: 1
   jobTemplate:
     spec:
+      # Retry the Job 3 times before marking it as failed.
+      backoffLimit: 3
+      # Delete the Job and its pods 24 hours after completion.
+      ttlSecondsAfterFinished: 86400
       template:
         metadata:
           annotations:

--- a/charts/murmurations/charts/nodecleaner/templates/nodecleaner/cronjob.yaml
+++ b/charts/murmurations/charts/nodecleaner/templates/nodecleaner/cronjob.yaml
@@ -8,8 +8,16 @@ spec:
   {{- else }}
   schedule: "*/2 * * * *"  # every 2 minutes
   {{- end }}
+  # Keep the latest successful job.
+  successfulJobsHistoryLimit: 1
+  # Keep the latest failed job.
+  failedJobsHistoryLimit: 1
   jobTemplate:
     spec:
+      # Retry the Job 3 times before marking it as failed.
+      backoffLimit: 3
+      # Delete the Job and its pods 24 hours after completion.
+      ttlSecondsAfterFinished: 86400
       template:
         metadata:
           annotations:

--- a/charts/murmurations/charts/revalidatenode/templates/revalidatenode/cronjob.yaml
+++ b/charts/murmurations/charts/revalidatenode/templates/revalidatenode/cronjob.yaml
@@ -8,8 +8,16 @@ spec:
   {{- else }}
   schedule: "*/2 * * * *"  # every two minutes
   {{- end }}
+  # Keep the latest successful job.
+  successfulJobsHistoryLimit: 1
+  # Keep the latest failed job.
+  failedJobsHistoryLimit: 1
   jobTemplate:
     spec:
+      # Retry the Job 3 times before marking it as failed.
+      backoffLimit: 3
+      # Delete the Job and its pods 24 hours after completion.
+      ttlSecondsAfterFinished: 86400
       template:
         metadata:
           annotations:

--- a/charts/murmurations/charts/schemaparser/templates/schemaparser/cronjob.yaml
+++ b/charts/murmurations/charts/schemaparser/templates/schemaparser/cronjob.yaml
@@ -8,8 +8,16 @@ spec:
   {{- else }}
   schedule: "*/2 * * * *"  # every two minutes
   {{- end }}
+  # Keep the latest successful job.
+  successfulJobsHistoryLimit: 1
+  # Keep the latest failed job.
+  failedJobsHistoryLimit: 1
   jobTemplate:
     spec:
+      # Retry the Job 3 times before marking it as failed.
+      backoffLimit: 3
+      # Delete the Job and its pods 24 hours after completion.
+      ttlSecondsAfterFinished: 86400
       template:
         metadata:
           annotations:

